File: leetcode/100-999/leetcode946.cpp
#include <algorithm>
#include <iostream>
#include <stack>
#include <vector>
using namespace std;
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped)
    {
        if (pushed.size() != popped.size())
            return false;
        stack<int> s;
        int j = 0;
        for (int i = 0; i < popped.size(); i++) {
            while ((s.empty() || popped[i] != s.top()) && j < pushed.size()) {
                s.push(pushed[j++]);
            }
            if (s.top() == popped[i])
                s.pop();
            else
                return false;
        }
        return true;
    }
};
int main(void) { return 0; }

File: leetcode/100-999/leetcode946.py
from typing import List


class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        st = []
        while popped:
            if not pushed and not st:
                return False
            if st and st[-1] == popped[0]:
                st.pop()
                popped.pop(0)
            else:
                if not pushed:
                    return False
                else:
                    st.append(pushed.pop(0))
        return True


s = Solution()
print(s.validateStackSequences(pushed=[1, 2, 3, 4, 5], popped=[4, 5, 3, 2, 1]))
